src/Orleans.Runtime/Utilities/AsyncEnumerable.cs

```diff
@@ -0,0 +1,241 @@
+
+                this.TryPublish(Element.CreateDisposed());
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowInvalidUpdate() => throw new ArgumentException("The value was not valid");
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowConcurrency() => throw new InvalidOperationException("An update was concurrently published by another thread");
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowDisposed() => throw new ObjectDisposedException("This instance has been disposed");
+
+        public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            return new AsyncEnumerator(this.current, cancellationToken);
+        }
+
+        private sealed class AsyncEnumerator : IAsyncEnumerator<T>
+        {
+            private readonly Task cancellation;
+            private Element current;
+
+            public AsyncEnumerator(Element initial, CancellationToken cancellation)
+            {
+                if (!initial.IsValid) this.current = initial;
+                else
+                {
+                    var result = Element.CreateInitial();
+                    result.SetNext(initial);
+                    this.current = result;
+                }
+
+                if (cancellation != default)
+                {
+                    this.cancellation = cancellation.WhenCancelled();
+                }
+            }
+
+            T IAsyncEnumerator<T>.Current => this.current.Value;
+
+            async ValueTask<bool> IAsyncEnumerator<T>.MoveNextAsync()
+            {
+                Task<Element> next;
+                if (this.cancellation != default)
+                {
+                    next = this.current.NextAsync();
+                    var result = await Task.WhenAny(this.cancellation, next);
+                    if (ReferenceEquals(result, this.cancellation)) return false;
+                }
+                else
+                {
+                    next = this.current.NextAsync();
+                }
+
+                this.current = await next;
+                return this.current.IsValid;
+            }
+
+            ValueTask IAsyncDisposable.DisposeAsync() => default;
+        }
+
+        private sealed class Element
+        {
+            private readonly TaskCompletionSource<Element> next;
+            private readonly object value;
+
+            public Element(T value)
+            {
+                this.value = value;
+                this.next = new TaskCompletionSource<Element>(TaskCreationOptions.RunContinuationsAsynchronously);
+            }
+
+            public static Element CreateInitial() => new Element(
+                AsyncEnumerable.InitialValue,
+                new TaskCompletionSource<Element>(TaskCreationOptions.RunContinuationsAsynchronously));
+
+            public static Element CreateDisposed()
```

Comment:
nit - Can't dispose element be an immutable static single instance.

Reply:
I think it's cleaner to create a new one each time. In theory there should be no difference

Reply:
Although I don't hold a strong opinion here